File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(vector<int> mats, vector<vector<string>> park) {
    int answer = -1;
    
    int R = park.size(), C = park[0].size();
    
    vector<vector<int>> dp(R, vector<int>(C, 0));
    
    int maxSq = 0;
    
    for (int i=0; i<R; i++) {
        for (int j=0; j<C; j++) {
            if (park[i][j] == "-1") {
                if (i==0 || j==0) {
                    dp[i][j] = 1;
                } else {
                    dp[i][j] = 1 + min({dp[i-1][j], dp[i][j-1], dp[i-1][j-1]});
                }
                maxSq = max(maxSq, dp[i][j]);
            }
        }
    }

    for (int m: mats) {
        if (m <= maxSq) {
            answer = max(answer, m);
        }
    }
    
    return answer;
}

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def solution(mats, park):
    R, C = len(park), len(park[0])
    answer = -1
    
    blank = []
    for i in range(R):
        for j in range(C):
            if park[i][j] == "-1":
                blank.append((i, j))
                
    for x, y in blank:
        for mat in mats:
            if x + mat > R or y + mat > C:
                continue

            possible = True
            for i in range(mat):
                for j in range(mat):
                    if park[x + i][y + j] != "-1":
                        possible = False
                        break
                if not possible:
                    break

            if possible:
                answer = max(answer, mat)
    
    return answer